src/main.rs

```diff
@@ -70,6 +92,116 @@ impl BFInterp {
             MetadataOperand => todo!(),
         }
     }
+
+    fn load(&mut self, op: &Operand, dest: &name::Name) {
+        match op {
+            LocalOperand { name, .. } => {
+                self.mem.insert(dest.clone(), *self.mem.get(name).unwrap());
+                self.mem.remove(name);
+            }
+            ConstantOperand(_) => todo!(),
+            MetadataOperand => todo!(),
+        }
+    }
+
+    fn call_func(&mut self, call: &Call) {
+        match &call.function {
+            Left(_) => todo!(),
+            Right(op) => match op {
+                LocalOperand { .. } => todo!(),
+                ConstantOperand(con_op) => match con_op.as_ref() {
+                    Constant::GlobalReference { name, .. } => match name {
+                        Name(n) => {
+                            if n.as_str() == "printf" {
+                                self.exe_printf(call);
+                            }
+                        }
+                        Number(_) => todo!(),
+                    },
+                    _ => todo!(),
+                },
+                MetadataOperand => todo!(),
+            },
+        }
+    }
+
+    fn exe_printf(&mut self, call: &Call) {
+        let constant = match &call.arguments[0].0 {
+            LocalOperand { .. } => todo!(),
+            ConstantOperand(const_op) => match const_op.as_ref() {
+                Constant::GetElementPtr(GetElementPtr { address, .. }) => match address.as_ref() {
+                    Constant::GlobalReference { name, .. } => self.gl_vars.get(name).unwrap(),
+                    _ => todo!(),
+                },
+                _ => todo!(),
+            },
+            MetadataOperand => todo!(),
+        };
+
+        let mut string = String::new();
+        let mut arg_it = call.arguments[1..].into_iter();
+        match constant {
+            Constant::Array { elements, .. } => {
+                let mut i = 0;
+                while i < elements.len() {
+                    match elements[i].as_ref() {
+                        Constant::Int { value, .. } => {
```

Comment:
I'm surprised this works, because according to the llvm-ir docs, it's `Constant::Int { bits, value }`, which means you should be reading out the bits here, not the value. But the tests seem to run. Any ideas why?

Reply:
The docs say that bits is the number of bits in the constant integer whereas value is the constant value itself. So the value is what we want to get the next char in the string.

Reply:
Yes, but when you do `Constant::Int { value, .. }`, you are saying, give me the first field of this enum (which is `bits`) and ignore all other fields (e.g. `value`). Using `value` in the match is just a name, it doesn't mean you are reading the `value` field, e.g. you could rename `value` to `v` here. I'm just very confused why the tests seem to work, as I believe they shouldn't, since you are reading the `bits` field here.

Reply:
I believe it is loading the value field as my ide is telling me that the type of value is &u64 and the docs say that the bits field is u32 and the value field is u64.

Reply:
Yes, I was wrong. Turns out that if you do `{ value, bits }` it looks up the fields by name, and the order doesn't matter (which in combination with `..` I personally find a bit confusing). You learn something new every day.